src/deh_main.c

```diff
@@ -0,0 +1,512 @@
+                    prev_section = NULL;
+                }
+
+                I_Printf(VB_DEBUG, "End '%s' tag", current_section->name);
+                current_section = NULL;
+            }
+        }
+        else
+        {
+            if (current_section != NULL)
+            {
+                // parse this line
+                current_section->line_parser(context, line, tag);
+            }
+            else
+            {
+                // possibly the start of a new section
+                char section_name[20];
+                sscanf(line, "%19s", section_name);
+                current_section = GetSectionByName(section_name);
+
+                if (current_section != NULL)
+                {
+                    I_Printf(VB_DEBUG, "Started '%s' tag", current_section->name);
+                    tag = current_section->start(context, line);
+                }
+                else if (prev_section != NULL)
+                {
+                    // [crispy] try this line again with the previous line parser
+                    DEH_RestoreLineStart(context);
+                    current_section = prev_section;
+                    prev_section = NULL;
+                }
+                else
+                {
+                    DEH_Warning(context, "Unkown DeHackEd section name '%s', WAD may not work properlly!", section_name);
+                }
+            }
+        }
+    }
+}
+
+// Parses a dehacked file
+int DEH_LoadFile(const char *filename)
+{
+    post_process = true;
+    if (!deh_initialized)
+    {
+        DEH_Init();
+    }
+
+    I_Printf(VB_DEBUG, "DEH_LoadFile: Loading %s", filename);
+    deh_context_t *context = DEH_OpenFile(filename);
+    if (context == NULL)
+    {
+        I_Printf(VB_ERROR, "DEH_LoadFile: Unable to open '%s'.", filename);
+        return 0;
+    }
+
+    AddDEHFileName(filename);
+
+    DEH_ParseContext(context);
+
+    DEH_CloseFile(context);
+
+    if (DEH_HadError(context))
+    {
+        I_Error("Error parsing dehacked file");
+    }
+
+    return 1;
+}
+
+// Load all dehacked patches from the given directory.
+void DEH_AutoLoadPatches(const char *path)
+{
+    const char *filename;
+    glob_t *glob = I_StartMultiGlob(path, GLOB_FLAG_NOCASE|GLOB_FLAG_SORTED, "*.deh", "*.bex");
+
+    for (;;)
+    {
+        filename = I_NextGlob(glob);
+        if (filename == NULL)
+        {
+            break;
+        }
+        DEH_LoadFile(filename);
+    }
+
+    I_EndGlob(glob);
+}
+
+// Load dehacked file from WAD lump.
+void DEH_LoadLump(int lumpnum)
+{
+    post_process = true;
+    if (!deh_initialized)
+    {
+        DEH_Init();
+    }
+
+    deh_context_t *context = DEH_OpenLump(lumpnum);
+    if (context == NULL)
+    {
+        I_Printf(VB_WARNING, "DEH_LoadFile: Unable to open lump %i", lumpnum);
+        return;
+    }
+
+    DEH_ParseContext(context);
+
+    DEH_CloseFile(context);
+
+    // If there was an error while parsing, abort.
+    if (DEH_HadError(context))
```

Comment:
Use after free. The `context` was freed in `DEH_CloseFile`.

Reply:
> is this really necessary?

I think so. The `deh_` code uses a mix of zone and standard malloc, which is annoying to support and debug. Back in the day, we decided to keep the Z_ functions in the renderer and playsim only. I will create a separate PR later.